compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -233,63 +228,66 @@ impl<K: DepKind> DepGraph<K> {
         cx: Ctxt,
         arg: A,
         task: fn(Ctxt, A) -> R,
-        create_task: fn(DepNode<K>) -> Option<TaskDeps<K>>,
-        hash_result: impl FnOnce(&mut Ctxt::StableHashingContext, &R) -> Option<Fingerprint>,
+        hash_result: fn(&mut Ctxt::StableHashingContext, &R) -> Option<Fingerprint>,
     ) -> (R, DepNodeIndex) {
-        if let Some(ref data) = self.data {
-            // If the following assertion triggers, it can have two reasons:
-            // 1. Something is wrong with DepNode creation, either here or
-            //    in `DepGraph::try_mark_green()`.
-            // 2. Two distinct query keys get mapped to the same `DepNode`
-            //    (see for example #48923).
-            assert!(
-                !self.dep_node_exists(&key),
-                "forcing query with already existing `DepNode`\n\
+        let data = self.data.as_ref().unwrap();
```

Comment:
The `unwrap()` can't fail since we're guarding this function in the calling code with `self.is_fully_enabled()`. Could you leave a comment to that effect?